File: LiftLog.Lib/Models/BlueprintModels.cs
using System.Collections.Immutable;

namespace LiftLog.Lib.Models;

public record ProgramBlueprint(
    string Name,
    ImmutableListValue<SessionBlueprint> Sessions,
    DateOnly LastEdited
);

public record SessionBlueprint(
    string Name,
    ImmutableListValue<ExerciseBlueprint> Exercises,
    string Notes
)
{
    public static readonly SessionBlueprint Empty = new(string.Empty, [], string.Empty);

    public Session GetEmptySession()
    {
        static RecordedExercise GetNextExercise(ExerciseBlueprint e)
        {
            return new RecordedExercise(
                e,
                Enumerable.Repeat(new PotentialSet(null, 0), e.Sets).ToImmutableList(),
                null,
                false
            );
        }

        return new Session(
            Guid.NewGuid(),
            this,
            Exercises.Select(GetNextExercise).ToImmutableList(),
            DateOnly.FromDateTime(DateTime.Now),
            null
        );
    }
}

public record ExerciseBlueprint(
    string Name,
    int Sets,
    int RepsPerSet,
    decimal WeightIncreaseOnSuccess,
    Rest RestBetweenSets,
    bool SupersetWithNext,
    string Notes,
    string Link
)
{
    public static readonly ExerciseBlueprint Default = new(
        Name: string.Empty,
        Sets: 3,
        RepsPerSet: 10,
        WeightIncreaseOnSuccess: 0,
        RestBetweenSets: Rest.Medium,
        SupersetWithNext: false,
        Notes: string.Empty,
        Link: string.Empty
    );
}

public record KeyedExerciseBlueprint(string Name, int Sets, int RepsPerSet)
{
    public static implicit operator KeyedExerciseBlueprint(ExerciseBlueprint e) =>
        new(e.Name, e.Sets, e.RepsPerSet);

    public class NormalizedNameOnlyEqualityComparer : IEqualityComparer<KeyedExerciseBlueprint>
    {
        public static readonly NormalizedNameOnlyEqualityComparer Instance = new();

        public bool Equals(KeyedExerciseBlueprint? x, KeyedExerciseBlueprint? y) =>
            NormalizeName(x?.Name) == NormalizeName(y?.Name);

        public int GetHashCode(KeyedExerciseBlueprint obj) => NormalizeName(obj.Name).GetHashCode();

        private static string NormalizeName(string? name)
        {
            if (name is null)
            {
                return string.Empty;
            }
            var lowerName = name.ToLower().Trim().Replace("flies", "flys").Replace("flyes", "flys");
            var withoutPlural = lowerName switch
            {
                string when lowerName.EndsWith("es") => lowerName[..^2],
                string when lowerName.EndsWith('s') => lowerName[..^1],
                _ => lowerName,
            };

            return withoutPlural;
        }
    }
}

public record Rest(TimeSpan MinRest, TimeSpan MaxRest, TimeSpan FailureRest)
{
    public static readonly Rest Short = new(
        TimeSpan.FromSeconds(60),
        TimeSpan.FromSeconds(90),
        TimeSpan.FromSeconds(180)
    );

    public static readonly Rest Medium = new(
        TimeSpan.FromSeconds(90),
        TimeSpan.FromSeconds(180),
        TimeSpan.FromSeconds(300)
    );

    public static readonly Rest Long = new(
        TimeSpan.FromMinutes(3),
        TimeSpan.FromMinutes(5),
        TimeSpan.FromMinutes(8)
    );
}

File: liftlog-react/models/blueprint-models.ts
import { Duration, LocalDate } from '@js-joda/core';
import BigNumber from 'bignumber.js';

export interface ProgramBlueprintPOJO {
  _BRAND: 'PROGRAM_BLUEPRINT_POJO';
  readonly name: string;
  readonly sessions: SessionBlueprintPOJO[];
  lastEdited: LocalDate;
}

export class ProgramBlueprint {
  readonly name: string;
  readonly sessions: SessionBlueprint[];
  lastEdited: LocalDate;

  /**
   * @deprecated please use full constructor. Here only for serialization
   */
  constructor();
  constructor(
    name: string,
    sessions: SessionBlueprint[],
    lastEdited: LocalDate,
  );
  constructor(
    name?: string,
    sessions?: SessionBlueprint[],
    lastEdited?: LocalDate,
  ) {
    this.name = name!;
    this.sessions = sessions!;
    this.lastEdited = lastEdited!;
  }

  static fromPOJO(
    pojo: Omit<ProgramBlueprintPOJO, '_BRAND'>,
  ): ProgramBlueprint {
    return new ProgramBlueprint(
      pojo.name,
      pojo.sessions.map(SessionBlueprint.fromPOJO),
      pojo.lastEdited,
    );
  }

  equals(other: ProgramBlueprint | undefined) {
    if (!other) {
      return false;
    }
    if (other === this) {
      return true;
    }

    return (
      this.name === other.name &&
      this.lastEdited.equals(other.lastEdited) &&
      this.sessions.length === other.sessions.length &&
      this.sessions.every((session, index) =>
        session.equals(other.sessions[index]),
      )
    );
  }

  toPOJO(): ProgramBlueprintPOJO {
    return {
      _BRAND: 'PROGRAM_BLUEPRINT_POJO',
      name: this.name,
      sessions: this.sessions.map((session) => session.toPOJO()),
      lastEdited: this.lastEdited,
    };
  }

  with(other: Partial<ProgramBlueprintPOJO>): ProgramBlueprint {
    return new ProgramBlueprint(
      other.name ?? this.name,
      other.sessions
        ? other.sessions.map(SessionBlueprint.fromPOJO)
        : this.sessions,
      other.lastEdited ?? this.lastEdited,
    );
  }
}

export interface SessionBlueprintPOJO {
  _BRAND: 'SESSION_BLUEPRINT_POJO';
  name: string;
  exercises: ExerciseBlueprintPOJO[];
  notes: string;
}

export class SessionBlueprint {
  readonly name: string;
  readonly exercises: ExerciseBlueprint[];
  readonly notes: string;

  /**
   * @deprecated please use full constructor. Here only for serialization
   */
  constructor();
  constructor(name: string, exercises: ExerciseBlueprint[], notes: string);
  constructor(name?: string, exercises?: ExerciseBlueprint[], notes?: string) {
    this.name = name!;
    this.exercises = exercises!;
    this.notes = notes!;
  }

  static fromPOJO(
    pojo: Omit<SessionBlueprintPOJO, '_BRAND'>,
  ): SessionBlueprint {
    return new SessionBlueprint(
      pojo.name,
      pojo.exercises.map(ExerciseBlueprint.fromPOJO),
      pojo.notes,
    );
  }

  equals(other: SessionBlueprint | undefined) {
    if (!other) {
      return false;
    }
    if (other === this) {
      return true;
    }

    return (
      this.name === other.name &&
      this.notes === other.notes &&
      this.exercises.length === other.exercises.length &&
      this.exercises.every((exercise, index) =>
        exercise.equals(other.exercises[index]),
      )
    );
  }

  toPOJO(): SessionBlueprintPOJO {
    return {
      _BRAND: 'SESSION_BLUEPRINT_POJO',
      name: this.name,
      exercises: this.exercises.map((exercise) => exercise.toPOJO()),
      notes: this.notes,
    };
  }

  with(other: Partial<SessionBlueprintPOJO>): SessionBlueprint {
    return new SessionBlueprint(
      other.name ?? this.name,
      other.exercises
        ? other.exercises.map(ExerciseBlueprint.fromPOJO)
        : this.exercises,
      other.notes ?? this.notes,
    );
  }
}

export interface ExerciseBlueprintPOJO {
  _BRAND: 'EXERCISE_BLUEPRINT_POJO';
  name: string;
  sets: number;
  repsPerSet: number;
  weightIncreaseOnSuccess: BigNumber;
  restBetweenSets: Rest;
  supersetWithNext: boolean;
  notes: string;
  link: string;
}

export class ExerciseBlueprint {
  readonly name: string;
  readonly sets: number;
  readonly repsPerSet: number;
  readonly weightIncreaseOnSuccess: BigNumber;
  readonly restBetweenSets: Rest;
  readonly supersetWithNext: boolean;
  readonly notes: string;
  readonly link: string;

  /**
   * @deprecated please use full constructor. Here only for serialization
   */
  constructor();
  constructor(
    name: string,
    sets: number,
    repsPerSet: number,
    weightIncreaseOnSuccess: BigNumber,
    restBetweenSets: Rest,
    supersetWithNext: boolean,
    notes: string,
    link: string,
  );
  constructor(
    name?: string,
    sets?: number,
    repsPerSet?: number,
    weightIncreaseOnSuccess?: BigNumber,
    restBetweenSets?: Rest,
    supersetWithNext?: boolean,
    notes?: string,
    link?: string,
  ) {
    this.name = name!;
    this.sets = sets!;
    this.repsPerSet = repsPerSet!;
    this.weightIncreaseOnSuccess = weightIncreaseOnSuccess!;
    this.restBetweenSets = restBetweenSets!;
    this.supersetWithNext = supersetWithNext!;
    this.notes = notes!;
    this.link = link!;
  }

  static fromPOJO(
    pojo: Omit<ExerciseBlueprintPOJO, '_BRAND'>,
  ): ExerciseBlueprint {
    return new ExerciseBlueprint(
      pojo.name,
      pojo.sets,
      pojo.repsPerSet,
      pojo.weightIncreaseOnSuccess,
      pojo.restBetweenSets,
      pojo.supersetWithNext,
      pojo.notes,
      pojo.link,
    );
  }

  equals(other: ExerciseBlueprint | undefined) {
    if (!other) {
      return false;
    }
    if (other === this) {
      return true;
    }

    return (
      this.name === other.name &&
      this.sets === other.sets &&
      this.repsPerSet === other.repsPerSet &&
      this.weightIncreaseOnSuccess.isEqualTo(other.weightIncreaseOnSuccess) &&
      this.restBetweenSets.minRest.equals(other.restBetweenSets.minRest) &&
      this.restBetweenSets.maxRest.equals(other.restBetweenSets.maxRest) &&
      this.restBetweenSets.failureRest.equals(
        other.restBetweenSets.failureRest,
      ) &&
      this.supersetWithNext === other.supersetWithNext &&
      this.notes === other.notes &&
      this.link === other.link
    );
  }

  toPOJO(): ExerciseBlueprintPOJO {
    return {
      _BRAND: 'EXERCISE_BLUEPRINT_POJO',
      name: this.name,
      sets: this.sets,
      repsPerSet: this.repsPerSet,
      weightIncreaseOnSuccess: this.weightIncreaseOnSuccess,
      restBetweenSets: this.restBetweenSets,
      supersetWithNext: this.supersetWithNext,
      notes: this.notes,
      link: this.link,
    };
  }

  with(other: Partial<ExerciseBlueprintPOJO>): ExerciseBlueprint {
    return new ExerciseBlueprint(
      other.name ?? this.name,
      other.sets ?? this.sets,
      other.repsPerSet ?? this.repsPerSet,
      other.weightIncreaseOnSuccess ?? this.weightIncreaseOnSuccess,
      other.restBetweenSets ?? this.restBetweenSets,
      other.supersetWithNext ?? this.supersetWithNext,
      other.notes ?? this.notes,
      other.link ?? this.link,
    );
  }
}

export class KeyedExerciseBlueprint {
  constructor(
    public name: string,
    public sets: number,
    public repsPerSet: number,
  ) {}

  static fromExerciseBlueprint(e: ExerciseBlueprint): KeyedExerciseBlueprint {
    return new KeyedExerciseBlueprint(e.name, e.sets, e.repsPerSet);
  }

  toString() {
    return `${this.name}_${this.sets}_${this.repsPerSet}`;
  }
}

export interface Rest {
  minRest: Duration;
  maxRest: Duration;
  failureRest: Duration;
}

export const Rest = {
  short: {
    minRest: Duration.ofSeconds(60),
    maxRest: Duration.ofSeconds(90),
    failureRest: Duration.ofSeconds(180),
  },
  medium: {
    minRest: Duration.ofSeconds(90),
    maxRest: Duration.ofSeconds(180),
    failureRest: Duration.ofSeconds(300),
  },
  long: {
    minRest: Duration.ofMinutes(3),
    maxRest: Duration.ofMinutes(5),
    failureRest: Duration.ofMinutes(8),
  },
} as const;
